app/src/org/commcare/android/nfc/NfcActivity.java

```diff
@@ -0,0 +1,150 @@
+            ((TextView)findViewById(R.id.nfc_instructions_text_view)).
+                    setText(Localization.get(getInstructionsTextKey()));
+        }
+    }
+
+    protected void initFields() {
+        this.nfcManager = new NfcManager(this);
+
+        this.domainForType = getIntent().getStringExtra(NFC_DOMAIN_ARG);
+        if (this.domainForType == null) {
+            // Usually there is no reason for the user to provide this themselves, and we will just
+            // set it to be the project's domain name
+            this.domainForType = ReportingUtils.getDomain();
+        }
+    }
+    /**
+     * Create an intent for restarting this activity, which will be passed to enableForegroundDispatch(),
+     * thus instructing Android to start the intent when the device detects a new NFC tag. Adding
+     * FLAG_ACTIVITY_SINGLE_TOP makes it so that onNewIntent() can be called in this activity when
+     * the intent is started.
+     **/
+    protected void createPendingRestartIntent() {
+        Intent i = new Intent(this, getClass());
+        i.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
+        this.pendingNfcIntent = PendingIntent.getActivity(this, 0, i, 0);
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
+            finishWithErrorToast("nfc.min.version.message");
+            return;
+        }
+
+        try {
+            nfcManager.checkForNFCSupport();
+            if (requiredFieldsMissing()) {
+                return;
+            }
+            setReadyToHandleTag();
+        } catch (NfcManager.NfcNotEnabledException e) {
+            finishWithErrorToast("nfc.not.enabled");
+        } catch (NfcManager.NfcNotSupportedException e) {
+            finishWithErrorToast("nfc.not.supported");
+        }
+    }
+
+    @Override
+    protected void onPause() {
+        super.onPause();
+        this.nfcManager.disableForegroundDispatch(this);
+    }
+
+    protected abstract boolean requiredFieldsMissing();
+
+    /**
+     * Makes it so that this activity will be the default to handle a new tag when it is discovered.
+     *
+     * The intent filters being passed to enableForegroundDispatch() here are intentionally very
+     * broad, on the assumption that if CommCare's NfcActivity is in the foreground while a user
+     * tries to scan an NFC tag, they were intending to scan something that CommCare would
+     * recognize. So if the user scans a tag of a type that we aren't expecting, instead of
+     * filtering it out (which would result in the device's default NFC handler consuming the tag,
+     * resulting in confusing on-screen behavior for the user), we'll consume it ourselves and
+     * then show a useful error message.
+     */
+    private void setReadyToHandleTag() {
+        IntentFilter ndefDiscoveredFilter = new IntentFilter(NfcAdapter.ACTION_NDEF_DISCOVERED);
```

Comment:
Is this not handled by the additions to the `AndroidManifest`? Or does this guarantee us higher priority?

Reply:
This is the setup Android outlines for making it so that an activity has highest priority for handling an NFC tag when, and _only_ when, that activity is in the foreground. So this way we only try to pick up NFC tags when the user has explicitly launched `NfcActivity` (either by clicking the Intent callout button in a callout question in a form, or by clicking the search callout button in a case list).